pkg/action/install.go

```diff
@@ -23,21 +44,116 @@ func NewInstall(cfg *Configuration) *Install {
 }
 
 // CheckDependencies checks the dependencies for a chart.
+// by wrapping action.CheckDependencies
 func CheckDependencies(ch *chart.Chart, reqs []*chart.Dependency) error {
-	var missing []string
+	return action.CheckDependencies(ch, reqs)
+}
 
-OUTER:
-	for _, r := range reqs {
-		for _, d := range ch.Dependencies() {
-			if d.Name() == r.Name {
-				continue OUTER
+// Run executes the installation
+//
+// If DryRun is set to true, this will prepare the release, but not install it
+func (i *Install) Run(chrt *chart.Chart, vals map[string]interface{}) (*release.Release, error) {
+	log.Infof("Installing chart \"%s\" in namespace \"%s\"…", i.ReleaseName, i.Namespace)
+	helmInstall := i.Install
+	rel, err := helmInstall.Run(chrt, vals) // wrap Helm's i.Run for now
+	return rel, err
+}
+
+// SetNamespace sets the Namespace that should be used in action.Install
+//
+// This will read the chart annotations. If no annotations, it leave the existing ns in the action.
+func (i *Install) SetNamespace(chart *chart.Chart, defaultns string) {
+	i.Namespace = defaultns
+	if chart.Metadata.Annotations != nil {
+		if val, ok := chart.Metadata.Annotations["hypper.cattle.io/namespace"]; ok {
+			i.Namespace = val
```

Comment:
This is one of the most confusing parts of the Helm API but this is not setting the namespace the kube client will put the object in. In `action.New` the namespace is set and this is the namespace that will end up being used to put the resources in. This is because `action.Init` creates a kubernetes client with a namespace set by a flag (in helm) or environment variable.

Reply:
Thanks, I see now! (debugger ftw). The complication I have is that I have no way of knowing if the namespace already set up in the `action` comes from the flag (first priority case) or comes from the kubeconfig context (last priority case). The namespace logic was always out of _action/install.go_, but now with annotations (which is in the middle of priority cases), we ought to provide a getter at least. So I suppose the logic between picking from the `action.Namespace` and the annotations may be out of the _action/install.go_ library.

Reply:
We fixed it in the last 4 comments.